spec.md

```diff
@@ -1617,6 +1642,53 @@ The CO MUST implement the specified error recovery behavior when it encounters t
 |-----------|-----------|-------------|-------------------|
 | Volume does not exist | 5 NOT_FOUND | Indicates that a volume corresponding to the specified `volume_id` does not exist. | Caller MUST verify that the `volume_id` is correct and that the volume is accessible and has not been deleted before retrying with exponential back off. |
 
+
+#### `ControllerModifyVolume`
+
+A Controller plugin MUST implement this RPC call if plugin has MODIFY_VOLUME controller capability.
+This RPC allows the CO to change mutable key attributes of a volume. 
+
+This operation MUST be idempotent. 
+The new mutable parameters in ControllerModifyVolume can be different from the existing mutable parameters. 
+
+
+```protobuf
+message ControllerModifyVolumeRequest {
+  option (alpha_message) = true;
+
+  // Contains identity information for the existing volume.
+  // This field is REQUIRED.
+  string volume_id = 1;
+
+  // Secrets required by plugin to complete modify volume request.
+  // This field is OPTIONAL. Refer to the `Secrets Requirements`
+  // section on how to use this field.
+  map<string, string> secrets = 2 [(csi_secret) = true];
+
+  // Plugin specific parameters to apply, passed in as opaque key-value
+  // pairs. This field is OPTIONAL. The Plugin is responsible for
+  // parsing and validating these parameters. COs will treat these
+  // as opaque. COs MUST specify the intended value of every mutable
+  // parameter. Absent keys that were previously assigned a value MUST
+  // be interpreted as a deletion of the key from the set of mutable
+  // parameters.
+  map<string, string> mutable_parameters = 3;
+}
+
+message ControllerModifyVolumeResponse {
+  option (alpha_message) = true;
+}
+
+```
+
+##### ControllerModifyVolume Errors
```

Comment:
i'm left wondering about mutable parameter changes that are not safe/possible to apply given the current state of the volume. e.g. the volume is node-published, and so tweaking parameter X is no longer possible in this state. however, it would be possible if the the volume were neither node-staged or node-published.

and if there are such cases, it would seem there's a need to talk about possible volume lifecycle dependencies somewhere (other than the error table for the RPC).

should there be a FAILED_PRECONDITION code for such cases? or else has this already been discussed?

Reply:
This is a good catch, and it hadn't been discussed before. All of us were focused on the kinds of mutations that would be safe to do under any circumstances, but you're right that one could imagine some that were only safe to do on a unpublished volume.

Would you be opposed to limiting this feature to the kind of mutations that were always safe to do for volumes regardless of their published state? That would remove any need to discuss lifecycle here. If we permit the plugin to return FAILED_PRECONDITION due to the volume lifecycle it's going to be very hard on the CO side to compensate. The CO implementation of this feature is likely to be a fairly simple reconciler loop that's blind to the publish state of the volume. The most likely outcome of allowing the SP to return that error is that some volume just won't reconcile until the user detaches them and a constant stream of errors in the mean time.

Reply:
It is probably imaginable that a change in parameters requires re-staging the volume to apply the parameters to a new connection (think of hostname change of an NFS-server). It may not be required immediately, but the change of the parameters only gets into effect once the volume has been disconnected and newly connected.

A `FAILED_PRECONDITION` is not sufficient in that case, I think. Maybe the `ControllerModifyVolumeResponse` could include an attribute to signal this?

Reply:
interesting proposal. something like..
```protobuf
  // when unset, communicates to the CO that one or more mutable parameters
  // require a volume lifecycle state change before becoming effective.
  string awaiting_state_transition = N;
```
.. which could be even tighter w/ an enum for the blocking state transition